primitives/src/merkle_tree/field_based_mht/path.rs

```diff
@@ -0,0 +1,380 @@
+use crate::{
+    crh::*, field_based_mht::*, Error
+};
+use std::{
+    clone::Clone, fmt::Debug, io::{Write, Result as IoResult, Read}
+};
+
+/// An implementation of the FieldBasedMerkleTreePath trait, for a given FieldBasedHash and
+/// FieldBasedMerkleTree with arbitrary arity.
+/// TODO: Test for arity > 2
+#[derive(Clone, Debug)]
+pub struct FieldBasedMHTPath<T: FieldBasedMerkleTreeParameters>{
+    path: Vec<(Vec<<T::H as FieldBasedHash>::Data>, usize)>,
+}
+
+impl<T: FieldBasedMerkleTreeParameters> FieldBasedMHTPath<T> {
+
+    /// Returns true if `self` is a Merkle Path for the left most leaf of a Merkle Tree,
+    /// false, otherwise.
+    #[inline]
+    pub fn is_leftmost(&self) -> bool {
+        for &(_, direction) in &self.path {
+            if direction != 0 {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /// Returns true if `self` is a Merkle Path for the right most leaf of a Merkle Tree,
+    /// false, otherwise.
+    #[inline]
+    pub fn is_rightmost(&self) -> bool {
+        for &(_, direction) in &self.path {
+            if direction != (T::MERKLE_ARITY - 1) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /// Returns true if `self` is a Merkle Path for the righmost non-empty leaf of the Merkle Tree
+    /// (e.g. the leaf which is not physically in the rightmost position of the tree, but it's
+    /// followed by all empty leaves).
+    /// Assumptions:
+    /// 1) Append-only Merkle Tree;
+    /// 2) T::EMPTY_HASH_CST is specified;
+    /// 3) Not to be called on Merkle Path corresponding to an empty leaf.
+    #[inline]
+    pub fn is_non_empty_rightmost(&self) -> bool {
```

Comment:
rename to are_right_leaves_empty, add it to trait and remove above assumptions

Reply:
Done